### PR TITLE
Make BFS correct again

### PR DESCRIPTION
**Pull request by leonardoalt**
The current BFS doesn't actually perform a BFS, it performs a mix of a BFS and an ordered search. This is confusing and could lead to problems.

### CODE REVIEW

**Comment by leonardoalt** [comment]
@ekpyron afaict all the modified places are interested in unordered reachability

File: libsolutil/Algorithms.h [hunk 1]
```diff
@@ -79,9 +79,6 @@ class CycleDetector
 /**
  * Generic breadth first search.
  *
- * Note that V needs to be a comparable value type. If it is not, use a pointer type,
- * but note that this might lead to non-deterministic traversal.
```
Comment (ekpyron) [inline libsolutil/Algorithms.h:83]: The first sentence is still true - you'll still need a comparison operator in order to be able to put stuff in ``visited``...

File: libyul/optimiser/Semantics.cpp [hunk 2]
```diff
@@ -125,21 +125,24 @@ map<YulString, SideEffects> SideEffectsPropagator::sideEffects(
...
-				{
-					if (ret.count(_function))
-						sideEffects += ret[_function];
-					for (YulString callee: _directCallGraph.functionCalls.at(_function))
-						_addChild(callee);
-				}
+		auto _visit = [&, visited = std::set<YulString>{}](YulString _function, auto&& _recurse) mutable {
+			if (visited.count(_function))
+				return;
+			visited.insert(_function);
```
Comment (ekpyron) [inline libyul/optimiser/Semantics.cpp]:
```suggestion
			if (!visited.insert(_function).second))
				return;
```
would actually be the same and would be faster - but it's a readability/speed tradeoff.

**Review by ekpyron — DISMISSED** [review]
Only minor comments. Generally LGTM.

File: libsolutil/Algorithms.h (hunk 1 above)
Comment (leonardoalt) [reply libsolutil/Algorithms.h:83]: true

File: libyul/optimiser/Semantics.cpp (hunk 2 above)
Comment (leonardoalt) [reply libyul/optimiser/Semantics.cpp]: Changed

**Comment by leonardoalt** [comment]
@ekpyron updated

**Review by ekpyron — APPROVED** [review]

File: libsolutil/Algorithms.h [hunk 3]
```diff
@@ -104,13 +103,14 @@ struct BreadthFirstSearch
 	{
 		while (!verticesToTraverse.empty())
 		{
-			V v = *verticesToTraverse.begin();
-			verticesToTraverse.erase(verticesToTraverse.begin());
+			V v = std::move(verticesToTraverse.front());
+			verticesToTraverse.pop_front();
+
 			visited.insert(v);
```
Comment (chriseth) [inline libsolutil/Algorithms.h:109]: We only mark a node as visited when we visit it. If multiple edges point to the same node, then this node can be inserted into the queue multiple times and thus will be visited multiple times. We should at least change this line to
```
if (!visited.insert(v).second)
  continue;
```
here.

Comment (chriseth) [reply libsolutil/Algorithms.h:109]: Another solution would be to change the `visited` set into a `queued` set that is set to true once the node has been seen in line 113

Comment (leonardoalt) [reply libsolutil/Algorithms.h:109]: It cannot be inserted into the queue multiple times exactly because we check before adding to the queue

Comment (leonardoalt) [reply libsolutil/Algorithms.h:109]: The only case where that's possible is if you initialize `verticesToVisit` with multiple copies of the same value

**Pull request merged**